File: Overlays/benofficial2 - iRacing Relative/JavascriptExtensions/Relative.js
/*
    benofficial2's Official Overlays
    Copyright (C) 2023-2025 benofficial2

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <https://www.gnu.org/licenses/>.
*/

const g_UseSimHubProp = false;

function getRelativeProp(index, prop)
{
    if (index > 0)
    {
        return $prop('benofficial2.Relative.Behind' + format(index - 1, '00') + '.' + prop);
    }
    return $prop('benofficial2.Relative.Ahead' + format(Math.abs(index + 1), '00') + '.' + prop);
}

function getStandingsClassProp(classIdx, prop)
{
    return $prop('benofficial2.Standings.Class' + format(classIdx, '00') 
        + '.' + prop);
}

function getRelativeRowVisible(index)
{
    if (index == 0 || g_UseSimHubProp)
    {
        const position = getopponentleaderboardposition_aheadbehind(index);
        return position >= 0;
    }
    else
    {
        return getRelativeProp(index, 'RowVisible');
    }
}

function getRelativePositionInClass(index)
{
    let posInClass = 0;
    if (g_UseSimHubProp)
    {
        const position = getopponentleaderboardposition_aheadbehind(index);
        posInClass = driverclassposition(position);
    }
    else if (index == 0)
    {
        posInClass = isnull($prop('benofficial2.Player.LivePositionInClass'), 0);
    }
    else
    {
        posInClass = getRelativeProp(index, 'LivePositionInClass');
    }
    return posInClass > 0 ? posInClass : '';
}

function getRelativeNumber(index)
{
    let number = '';
    if (index == 0 || g_UseSimHubProp)
    {
        const position = getopponentleaderboardposition_aheadbehind(index);
        number = drivercarnumber(position);
    }
    else
    {
        number = getRelativeProp(index, 'Number');
    }
    return '#' + number;
}

function getRelativeClassColor(index)
{
    if (index == 0 || g_UseSimHubProp)
    {
        const position = getopponentleaderboardposition_aheadbehind(index);
        return drivercarclasscolor(position);
    }
    else
    {
        return getRelativeProp(index, 'ClassColor');
    }
}

function getRelativeIsOutLap(index)
{
    if (g_UseSimHubProp)
    {
        const position = getopponentleaderboardposition_aheadbehind(index);
        return driverisoutlap(position);
    }
    else if (index == 0)
    {
        return isnull($prop('benofficial2.Player.OutLap'), false);
    }
    else
    {
        return getRelativeProp(index, 'OutLap');
    }
}

function getRelativeSessionFlags(index)
{
    if (index == 0)
    {
        return isnull($prop('DataCorePlugin.GameRawData.Telemetry.SessionFlags'), 0);
    }
    else
    {
        return getRelativeProp(index, 'SessionFlags');
    }
}

function getRelativeName(index)
{
    if (index == 0 || g_UseSimHubProp)
    {
        const position = getopponentleaderboardposition_aheadbehind(index);
        return drivername(position);
    }
    else
    {
        return getRelativeProp(index, 'Name');
    }
}

function getRelativeCarBrand(index)
{
    if (index == 0)
    {
        return $prop('benofficial2.Player.CarBrand');
    }
    else
    {
        return getRelativeProp(index, 'CarBrand');
    }
}

function getRelativeCountryCode(index)
{
    if (index == 0)
    {
        return $prop('benofficial2.Player.CountryCode');
    }
    else
    {
        return getRelativeProp(index, 'CountryCode');
    }
}

function getRelativeIRating(index)
{
    let iRating;
    if (index == 0 || g_UseSimHubProp)
    {
        const position = getopponentleaderboardposition_aheadbehind(index);
        iRating = driveriracingirating(position);
    }
    else
    {
        iRating = getRelativeProp(index, 'iRating');
    }
    return formatIRating(iRating);
}

function getRelativeIRatingChange(index)
{
    if (index == 0)
    {
        return isnull($prop('benofficial2.Player.iRatingChange'), 0)
    }
    else
    {
        return isnull(getRelativeProp(index, 'iRatingChange'), 0)
    }
}

function getRelativeSafetyRating(index)
{
    let iRating;
    if (index == 0 || g_UseSimHubProp)
    {
        const position = getopponentleaderboardposition_aheadbehind(index);
        return Number(String(driverlicencestring(position)).slice(2, 6));
    }
    else
    {
        return getRelativeProp(index, 'SafetyRating');
    }
}

function getRelativeLicense(index)
{
    if (index == 0 || g_UseSimHubProp)
    {
        const position = getopponentleaderboardposition_aheadbehind(index);
        return String(driverlicencestring(position)).slice(0, -5);
    }
    else
    {
        return getRelativeProp(index, 'License');
    }
}

function getRelativeLicenseColor(index)
{
    const license = getRelativeLicense(index);
    if (license == 'A')
    {
        return '#006EFF'
    }
    else if (license == 'B')
    {
        return '#33CC00'
    }
    else if (license == 'C')
    {
        return '#FFCC00'
    }
    else if (license == 'D')
    {
        return '#FF6600'
    }
    else if (license == 'R')
    {
        return '#E1251B'
    }
    return '#808085'
}

function getRelativeLicenseTextColor(index)
{
    const license = getRelativeLicense(index);
    if (license == 'A')
    {
        return '#66A8FF'
    }
    else if (license == 'B')
    {
        return '#85E066'
    }
    else if (license == 'C')
    {
        return '#FFE066'
    }
    else if (license == 'D')
    {
        return '#FFA366'
    }
    else if (license == 'R')
    {
        return '#ED7C66'
    }
    return '#AEAEB0'
}

function getRelativeLicenseBackColor(index)
{
    const license = getRelativeLicense(index);
    if (license == 'A')
    {
        return '#032F6F'
    }
    else if (license == 'B')
    {
        return '#175509'
    }
    else if (license == 'C')
    {
        return '#50410A'
    }
    else if (license == 'D')
    {
        return '#692C09'
    }
    else if (license == 'R')
    {
        return '#5D1214'
    }
    return '#37373F'
}

function getRelativeGapToPlayer(index)
{
    let gap;
    if (index == 0)
    {
        return '';
    }
    else if (g_UseSimHubProp)
    {
        const position = getopponentleaderboardposition_aheadbehind(index);
        gap = driverrelativegaptoplayer(position);
    }
    else
    {
        gap = getRelativeProp(index, 'GapToPlayer');
    }
    return Math.min(999.9, Math.abs(gap).toFixed(1));
}

function getRelativeTextColor(index)
{
    if (!isRace()) return 'White';

    if (index == 0)
    {
        const highlight = isnull($prop('benofficial2.Relative.HighlightPlayerRow'), false)
        if (highlight) return 'White'
        return '#FFEBAE00'
    }

    let lap;
    if (g_UseSimHubProp)
    {
        const position = getopponentleaderboardposition_aheadbehind(index);
        lap = drivercurrentlaphighprecision(position);
    }
    else
    {
        lap = getRelativeProp(index, 'CurrentLapHighPrecision');
    }

    const playerPosition = getopponentleaderboardposition_aheadbehind(0);
    const playerLap = drivercurrentlaphighprecision(playerPosition);

    if (playerLap <= 0 || lap <= 0) return 'White';

    const red = '#FFFF6345';
    const blue = '#43B7EA';
    const white = 'White';

    let deltaLap = playerLap - lap;

    if (index < 0)
    {
        if (deltaLap < -0.75) return red;
        if (deltaLap > 0.25) return blue;
    }
    else if (index > 0)
    {
        if (deltaLap < -0.25) return red;
        if (deltaLap > 0.75) return blue;
    }
    return white;
}

function getClassSof(classIdx)
{
    return isnull(getStandingsClassProp(classIdx, 'Sof'), 0);
}

function truncateToDecimal(num, decimals) 
{
  const factor = Math.pow(10, decimals)
  return (Math.floor(num * factor) / factor).toFixed(decimals)
}

function formatIRating(iRating)
{
    return truncateToDecimal(Number(iRating) / 1000, 1) + 'k';
}

File: Plugin/Modules/Relative.cs
/*
    benofficial2's Official Overlays
    Copyright (C) 2025 benofficial2

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <https://www.gnu.org/licenses/>.
*/

using GameReaderCommon;
using SimHub.Plugins;
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Linq;
using System.Security.Policy;

namespace benofficial2.Plugin
{
    public class RelativeSettings : ModuleSettings
    {
        public int MaxRows { get; set; } = 3;
        public bool HeaderVisible { get; set; } = true;
        public int HeaderOpacity { get; set; } = 90;
        public bool CarLogoVisible { get; set; } = true;
        public int AlternateRowBackgroundColor { get; set; } = 5;
        public bool HighlightPlayerRow { get; set; } = true;
        public int BackgroundOpacity { get; set; } = 60;
    }

    public class RelativeRow
    {
        public bool RowVisible { get; set; } = false;
        public int LivePositionInClass { get; set; } = 0;
        public string ClassColor { get; set; } = string.Empty;
        public string ClassTextColor { get; set; } = string.Empty;
        public string Number { get; set; } = string.Empty;
        public string Name { get; set; } = string.Empty;
        public string CarBrand { get; set; } = string.Empty;
        public bool OutLap { get; set; } = false;
        public int iRating { get; set; } = 0;
        public string License { get; set; } = string.Empty;
        public double SafetyRating { get; set; } = 0;
        public double GapToPlayer { get; set; } = 0;
        public string GapToPlayerCombined { get; set; } = string.Empty;
        public double CurrentLapHighPrecision { get; set; } = 0;
        public TimeSpan LastLapTime { get; set; } = TimeSpan.Zero;
    }

    public class RelativeAhead
    {
        public const int MaxRows = 5;
        public List<RelativeRow> Rows { get; internal set; }

        public RelativeAhead()
        {
            Rows = new List<RelativeRow>(Enumerable.Range(0, MaxRows).Select(x => new RelativeRow()));
        }
    }
    public class RelativeBehind
    {
        public const int MaxRows = 5;
        public List<RelativeRow> Rows { get; internal set; }

        public RelativeBehind()
        {
            Rows = new List<RelativeRow>(Enumerable.Range(0, MaxRows).Select(x => new RelativeRow()));
        }
    }

    public class RelativeModule : PluginModuleBase
    {
        private DateTime _lastUpdateTime = DateTime.MinValue;
        private TimeSpan _updateInterval = TimeSpan.FromMilliseconds(500);
        private DriverModule _driverModule = null;
        private CarModule _carModule = null;

        public RelativeSettings Settings { get; set; }

        public RelativeAhead Ahead = new RelativeAhead();
        public RelativeBehind Behind = new RelativeBehind();

        public override void Init(PluginManager pluginManager, benofficial2 plugin)
        {
            _driverModule = plugin.GetModule<DriverModule>();
            _carModule = plugin.GetModule<CarModule>();

            Settings = plugin.ReadCommonSettings<RelativeSettings>("RelativeSettings", () => new RelativeSettings());
            plugin.AttachDelegate(name: "Relative.MaxRows", valueProvider: () => Settings.MaxRows);
            plugin.AttachDelegate(name: "Relative.HeaderVisible", valueProvider: () => Settings.HeaderVisible);
            plugin.AttachDelegate(name: "Relative.HeaderOpacity", valueProvider: () => Settings.HeaderOpacity);
            plugin.AttachDelegate(name: "Relative.CarLogoVisible", valueProvider: () => Settings.CarLogoVisible);
            plugin.AttachDelegate(name: "Relative.AlternateRowBackgroundColor", valueProvider: () => Settings.AlternateRowBackgroundColor);
            plugin.AttachDelegate(name: "Relative.HighlightPlayerRow", valueProvider: () => Settings.HighlightPlayerRow);
            plugin.AttachDelegate(name: "Relative.BackgroundOpacity", valueProvider: () => Settings.BackgroundOpacity);

            InitRelative(plugin, "Ahead", Ahead.Rows, RelativeAhead.MaxRows);
            InitRelative(plugin, "Behind", Behind.Rows, RelativeBehind.MaxRows);
        }

        private void InitRelative(benofficial2 plugin, string aheadBehind, List<RelativeRow> rows, int maxRows)
        {
            for (int rowIdx = 0; rowIdx < maxRows; rowIdx++)
            {
                RelativeRow row = rows[rowIdx];
                plugin.AttachDelegate(name: $"Relative.{aheadBehind}{rowIdx:00}.RowVisible", valueProvider: () => row.RowVisible);
                plugin.AttachDelegate(name: $"Relative.{aheadBehind}{rowIdx:00}.LivePositionInClass", valueProvider: () => row.LivePositionInClass);
                plugin.AttachDelegate(name: $"Relative.{aheadBehind}{rowIdx:00}.ClassColor", valueProvider: () => row.ClassColor);
                plugin.AttachDelegate(name: $"Relative.{aheadBehind}{rowIdx:00}.ClassTextColor", valueProvider: () => row.ClassTextColor);
                plugin.AttachDelegate(name: $"Relative.{aheadBehind}{rowIdx:00}.Number", valueProvider: () => row.Number);
                plugin.AttachDelegate(name: $"Relative.{aheadBehind}{rowIdx:00}.Name", valueProvider: () => row.Name);
                plugin.AttachDelegate(name: $"Relative.{aheadBehind}{rowIdx:00}.CarBrand", valueProvider: () => row.CarBrand);
                plugin.AttachDelegate(name: $"Relative.{aheadBehind}{rowIdx:00}.OutLap", valueProvider: () => row.OutLap);
                plugin.AttachDelegate(name: $"Relative.{aheadBehind}{rowIdx:00}.iRating", valueProvider: () => row.iRating);
                plugin.AttachDelegate(name: $"Relative.{aheadBehind}{rowIdx:00}.License", valueProvider: () => row.License);
                plugin.AttachDelegate(name: $"Relative.{aheadBehind}{rowIdx:00}.SafetyRating", valueProvider: () => row.SafetyRating);
                plugin.AttachDelegate(name: $"Relative.{aheadBehind}{rowIdx:00}.GapToPlayer", valueProvider: () => row.GapToPlayer);
                plugin.AttachDelegate(name: $"Relative.{aheadBehind}{rowIdx:00}.GapToPlayerCombined", valueProvider: () => row.GapToPlayerCombined);
                plugin.AttachDelegate(name: $"Relative.{aheadBehind}{rowIdx:00}.CurrentLapHighPrecision", valueProvider: () => row.CurrentLapHighPrecision);
                plugin.AttachDelegate(name: $"Relative.{aheadBehind}{rowIdx:00}.LastLapTime", valueProvider: () => row.LastLapTime);
            }
        }

        public override void DataUpdate(PluginManager pluginManager, benofficial2 plugin, ref GameData data)
        {
            if (data.FrameTime - _lastUpdateTime < _updateInterval) return;
            _lastUpdateTime = data.FrameTime;

            UpdateRelative(ref data, Ahead.Rows, RelativeAhead.MaxRows, data.NewData.OpponentsAheadOnTrack);
            UpdateRelative(ref data, Behind.Rows, RelativeBehind.MaxRows, data.NewData.OpponentsBehindOnTrack);
        }

        public void UpdateRelative(ref GameData data, List<RelativeRow> rows, int maxRows, List<Opponent> opponents)
        {
            for (int rowIdx = 0; rowIdx < maxRows; rowIdx++)
            {
                RelativeRow row = rows[rowIdx];

                if (rowIdx >= opponents.Count)
                {
                    BlankRow(row);
                    continue;
                }

                Opponent opponent = opponents[rowIdx];
                Driver driver = null;
                if (_driverModule.Drivers != null) _driverModule.Drivers.TryGetValue(opponent.CarNumber, out driver);

                if (driver == null || !IsValidRow(opponent))
                {
                    BlankRow(row);
                    continue;
                }

                row.RowVisible = true;
                row.LivePositionInClass = driver.LivePositionInClass;
                row.ClassColor = opponent.CarClassColor;
                row.ClassTextColor = opponent.CarClassTextColor;
                row.Number = opponent.CarNumber;
                row.Name = opponent.Name;
                row.CarBrand = _carModule.GetCarBrand(driver.CarId, opponent.CarName); ;
                row.OutLap = driver.OutLap;
                row.iRating = (int)(opponent.IRacing_IRating ?? 0);
                (row.License, row.SafetyRating) = DriverModule.ParseLicenseString(opponent.LicenceString);
                row.GapToPlayer = opponent.RelativeGapToPlayer ?? 0;
                row.GapToPlayerCombined = opponent.GapToPlayerCombined;
                row.CurrentLapHighPrecision = opponent.CurrentLapHighPrecision ?? 0;
                row.LastLapTime = driver.LastLapTime;
            }
        }

        public override void End(PluginManager pluginManager, benofficial2 plugin)
        {
            plugin.SaveCommonSettings("RelativeSettings", Settings);
        }

        public void BlankRow(RelativeRow row)
        {
            row.RowVisible = false;
            row.LivePositionInClass = 0;
            row.ClassColor = string.Empty;
            row.ClassTextColor = string.Empty;
            row.Number = string.Empty;
            row.Name = string.Empty;
            row.CarBrand = string.Empty;
            row.OutLap = false;
            row.iRating = 0;
            row.License = string.Empty;
            row.SafetyRating = 0;
            row.GapToPlayer = 0;
            row.GapToPlayerCombined = string.Empty;
            row.CurrentLapHighPrecision = 0;
            row.LastLapTime = TimeSpan.Zero;
        }
        public bool IsValidRow(Opponent opponent)
        {
            return true;
        }
    }
}
